### DISCUSSION TITLE
使用folia开发插件的时候 并发操作我的数据 如何保证线程安全

### DISCUSSION THREAD

**Discussion by t13max** [mouse0w0/MinecraftDeveloperGuide#88] [问答 / Q&A]
folia是多线程tick的 我的某个数据是被大家共享的 可能会并发修改 如果加锁 会使问题复杂化 因为可能一次操作会涉及多个业务 修改多条数据 排序加锁也很复杂 使用任务队列的形式可以避免加锁 但是有些调用又是同步的 使用CompletableFuture的话会让folia的线程阻塞等待 玩家会感觉很卡 有什么好的解决方案吗 提前谢谢大家了

**Comment by t13max** [comment]
目前想的是可以先正常返回 等处理完在做一些后续的通知之类的 不知道能不能行 会不会有些业务不适用

**Comment by mouse0w0** [comment]
如果数据访问的频率不高，可以使用原子类型来存储值，并使用`compareAndSet`来设置值。但这可能是一个X-Y问题，最终还是需要按照实际的场景来分析取舍。

**Reply by t13max** [reply]
可是 我的数据对象会有很多字段 不可能每个都改成原子类型的 我感觉这不是X-Y问题 这种情况很常见 比如公会 公会中的多个玩家访问这个公会的数据 但是这些玩家是在不同的线程

**Reply by t13max** [reply]
大佬鼠 别直面天命了 不帮助群友 甚至今天又只修改了个README 太糊弄猫了

**Reply by mouse0w0** [reply]
从Folia官方文档来看，对于这种全局的数据修改，可以采用类似于[Global region task queue](https://docs.papermc.io/folia/reference/overview#global-region-task-queue)的方式去处理。本质上来说就是由一个线程专门的处理这些操作。

**Reply by t13max** [reply]
谢谢大佬**鼠**和群友们的帮助

**Reply by xtexx** [reply]
大老鼠草

**Reply by t13max** [reply]
不许胡说! 是大佬! **大佬鼠!**!!
